### PR TITLE
test different ids tensor to eliminate special separation/merge tokens

### PR DESCRIPTION
**Pull request by jeffjacobsen**
This was an experiment to try to remove the need for the separation token by creating the ids tensor once before the loop.  I'm not too familiar with tensors, so relied on ChatGPT to write the merge code.  There may be room for improvement, but the code seems to work with a single GPU or 2 GPUs.

### CODE REVIEW

**Comment by evintunador** [comment]
thanks Jeff! unfortunately I found an edge-case bug that I don't see any way of getting around. Basically whenever you're merging the final pair of a word and then the first token of the next word just so happens to match up with the final token of the prior word the script has no way of knowing that there's a regex barrier there and improperly replaces a token. here's some code to demonstrate
```Python
import torch

# works fine
#ids_lofl = [[1,2,3,4,1,2,2,4,1],[1],[4,3,2,1,4,3],[1,2],[3,4,2,1,3,2]]
# breaks at the boundary between the final two words (notice I changed 3 to 2, and 2 matches the [1,2] pair)
ids_lofl = [[1,2,3,4,1,2,2,4,1],[1],[4,3,2,1,4,3],[1,2],[2,4,2,1,3,2]]
print(ids_lofl)

# turn data into tensor pairs
ids1, ids2 = [], []
for word in ids_lofl:
    if len(word) > 1:
        ids1.extend(word[:-1])
        ids2.extend(word[1:])
pairs = torch.tensor([ids1, ids2])
print(pairs)

unique, counts = torch.unique(pairs, return_counts=True, dim=1)
print(unique, counts)

pair_idx = torch.argmax(counts) # (1)
best_pair = unique[:, pair_idx].cpu().numpy() # (2)
print(best_pair)

# Now merge that most common pair in all the words
match_mask = (pairs[0] == best_pair[0]) & (pairs[1] == best_pair[1]) 
# replace pairs[0] == best_pair[1] *after* the matched pairs
#after_mask = torch.zeros_like(match_mask)
#after_mask[1:] = match_mask[:-1]
after_mask = torch.roll(match_mask, 1)
replace_pairs_top_mask = after_mask & (pairs[0] == best_pair[1])
pairs[0][replace_pairs_top_mask] = 5
print(pairs)
# replace pairs[1] == best_pair[0] *before* the matched pairs
before_mask = torch.zeros_like(match_mask)
before_mask[:-1] = match_mask[1:]
replace_pairs_bot_mask = before_mask & (pairs[1] == best_pair[0])
pairs[1][replace_pairs_bot_mask] = 5
print(pairs)
# Delete the matched pairs
pairs = pairs[:, ~match_mask]
print(pairs)
```
unfortunate, as I was very excited for this. definitely a clever attempt but that edge case is killer. the separator tokens were providing that information between word boundaries; maybe you could maintain that information using a separate mask or something? but then that'd be an entire extra tensor so likely not worth it, idk. lmk if you see a fix, otherwise I'll close the pr

**Comment by jeffjacobsen** [comment]
That was what I was afraid of when I abandoned the idea the first time.  It
seemed like a good way to detect pairs but I felt like merging was going to
fail somehow.

There may be some solution with another mask, but agree that any speed
advantage may then be lost.

Go ahead and close the pr.  I may experiment with the mask idea just to
gain more experience with tensors, but I'm not holding my breath for any
great solution...
